sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -356,6 +367,71 @@ abstract class TypeCoercionBase {
         plan
       }
     }
+
+    /** Builds new children for set operators while resolving collation conflicts in respective
+     * children columns. This is being done before any other type widening because STRING can also
+     * be casted to a different dataType, but conflicts between collations have to be reported. */
+    private def buildNewChildrenWithResolvedCollation(
+      children: Seq[LogicalPlan]): Seq[LogicalPlan] = {
+      require(children.forall(_.output.length == children.head.output.length))
+      // Since this function is called for set operators, we can be sure that their children
+      // are all `Project` nodes. An `assert` is added for double checking the assumption.
```

Comment:
I'm confused, how do we guarantee it? The children of `Union` can be anything, right?